modules/cluster-management/src/main/java/org/apache/ignite/internal/cluster/management/ClusterManagementGroupManager.java

```diff
@@ -133,90 +144,384 @@ public void initCluster(Collection<String> metaStorageNodeNames, Collection<Stri
+                        .shouldCancel(!(e instanceof IllegalInitArgumentException))
+                        .build();
+            }
+
+            clusterService.messagingService().respond(addr, response, correlationId);
+        });
+    }
+
+    /**
+     * Starts the CMG Raft service and writes the given {@code state} to the storage.
+     */
+    private CompletableFuture<CmgRaftService> initCmgRaftService(ClusterState state) {
+        return localStateStorage.putCmgNodeNames(state.cmgNodes())
+                .thenCompose(v -> startCmgRaftService(state.cmgNodes()))
+                .thenCompose(service -> service.isCurrentNodeLeader()
+                        .thenCompose(isLeader -> isLeader ? initCmgState(service, state) : completedFuture(null))
+                        .thenApply(v -> service));
+    }
 
-                        if (leader.equals(thisNode)) {
-                            broadcastClusterState(msg.metaStorageNodes());
+    /**
+     * Writes the given state to the CMG's STM and executes some necessary on-leader logic.
+     */
+    private CompletableFuture<Void> initCmgState(CmgRaftService service, ClusterState state) {
+        return service.writeClusterState(state).thenCompose(v -> onLeaderElected(service, state));
+    }
+
+    /**
+     * Executes the following actions when a CMG leader is elected.
+     * <ol>
+     *     <li>Updates the logical topology in case some nodes have gone offline during leader election.</li>
+     *     <li>Broadcasts the current CMG state to all nodes in the physical topology.</li>
+     * </ol>
+     */
+    private CompletableFuture<Void> onLeaderElected(CmgRaftService service, ClusterState state) {
+        return updateLogicalTopology(service)
+                .thenRun(() -> {
+                    clusterService.topologyService().addEventHandler(cmgLeaderTopologyEventHandler(service));
+
+                    sendClusterState(state, clusterService.topologyService().allMembers());
+                });
+    }
+
+    /**
+     * This method must be executed upon CMG leader election in order to regain logical topology consistency in case some nodes left the
+     * physical topology during the election. New node will be added automatically after the new leader broadcasts the current cluster
+     * state.
+     */
+    private CompletableFuture<Void> updateLogicalTopology(CmgRaftService service) {
+        return service.logicalTopology()
+                .thenAccept(logicalTopology -> {
+                    Collection<String> physicalTopologyIds = clusterService.topologyService().allMembers()
+                            .stream()
+                            .map(ClusterNode::id)
+                            .collect(toSet());
+
+                    for (ClusterNode node : logicalTopology) {
+                        if (!physicalTopologyIds.contains(node.id())) {
+                            scheduleRemoveFromLogicalTopology(service, node);
```

Comment:
Simply for code reuse, do you think it may pose any problems?

Reply:
For me it seems to be totally unnecessary, why we need to schedule such removal? Could you please explain what kind of code reuse do you mean here?

Reply:
fixed, it was indeed unnecessary

Reply:
Let's update `service.removeFromCluster(node);` a bit and corresponding NodeLeaveCommand  to Node**s**LeaveCommand with 
    `private final Set<ClusterNode> node;`
instead of
   ` private final ClusterNode node;`
in order to remove all nodes at ones.

Reply:
done